src/ps.rs

```diff
@@ -0,0 +1,625 @@
+    Ok(js_sys::Uint8Array::from(bytes.as_slice()))
+}
+
+#[wasm_bindgen(js_name = psChallengeMessagesPoKContributionFromProtocol)]
+pub fn ps_challenge_messages_pok_contribution_from_protocol(
+    protocol: JsValue,
+    params: JsValue,
+    h: js_sys::Uint8Array,
+) -> Result<js_sys::Uint8Array, JsValue> {
+    set_panic_hook();
+
+    let protocol: PoKOfMessagesProtocol = from_value(protocol)?;
+    let params: SignatureParams = from_value(params)?;
+    let h = obj_from_uint8array!(G1Affine, h, false);
+    let mut bytes = vec![];
+
+    protocol
+        .challenge_contribution(&mut bytes, &params, &h)
+        .map_err(|e| {
+            JsValue::from(&format!(
+                "Evaluating challenge_contribution returned error: {:?}",
+                e
+            ))
+        })?;
+    Ok(js_sys::Uint8Array::from(bytes.as_slice()))
+}
+
+#[wasm_bindgen(js_name = psChallengeSignaturePoKContributionFromProof)]
+pub fn ps_challenge_signature_pok_contribution_from_proof(
+    proof: js_sys::Uint8Array,
+    public_key: js_sys::Uint8Array,
+    params: JsValue,
+) -> Result<js_sys::Uint8Array, JsValue> {
+    set_panic_hook();
+
+    let proof: PoKOfSignatureProof = obj_from_uint8array!(PoKOfSignatureProof, proof, false);
+    let pk = obj_from_uint8array!(PsPublicKey, public_key, false, "PsPublicKey");
+    let params: SignatureParams = from_value(params)?;
+
+    let mut bytes = vec![];
+    proof
+        .challenge_contribution(&mut bytes, &pk, &params)
+        .map_err(|e| {
+            JsValue::from(&format!(
+                "Evaluating challenge_contribution returned error: {:?}",
+                e
+            ))
+        })?;
+
+    Ok(js_sys::Uint8Array::from(bytes.as_slice()))
+}
+
+#[wasm_bindgen(js_name = psChallengeMessagesPoKContributionFromProof)]
+pub fn ps_challenge_messages_pok_contribution_from_proof(
+    proof: js_sys::Uint8Array,
+    params: JsValue,
+    h: js_sys::Uint8Array,
+) -> Result<js_sys::Uint8Array, JsValue> {
+    set_panic_hook();
+
+    let proof: PoKOfMessagesProof = obj_from_uint8array!(PoKOfMessagesProof, proof, false);
+    let params: SignatureParams = from_value(params)?;
+    let h = obj_from_uint8array!(G1Affine, h, false);
+
+    let mut bytes = vec![];
+    proof
+        .challenge_contribution(&mut bytes, &params, &h)
+        .map_err(|e| {
+            JsValue::from(&format!(
+                "Evaluating challenge_contribution returned error: {:?}",
+                e
+            ))
+        })?;
+
+    Ok(js_sys::Uint8Array::from(bytes.as_slice()))
+}
+
+#[wasm_bindgen(js_name = psSignatureParamsFromBytes)]
+pub fn ps_signature_params_from_bytes(bytes: js_sys::Uint8Array) -> Result<JsValue, JsValue> {
+    set_panic_hook();
+    let params = obj_from_uint8array!(SignatureParams, bytes, false, "SignatureParams");
+    serde_wasm_bindgen::to_value(&params).map_err(JsValue::from)
+}
+
+#[wasm_bindgen(js_name = psAdaptSignatureParamsForMsgCount)]
+pub fn adapt_sig_params_for_msg_count(
```

Comment:
This alone won't be sufficient because unlike in BBS+, in PS the public key need to have as many elements as the messages, and each such element can only be created with the knowledge of the secret key

Reply:
So what's the proper solution?
As for now, I use this function to adapt params and then recreate the private key from these params, and the public key can be built later using combined adapted params and secret key like [here](https://github.com/docknetwork/crypto-wasm-ts/pull/17/files#diff-d59cd0c4cabc1ecbccffb63366bac0c0ec3f9543ac40e67ac45b7b1d287e0ba6R210)

Reply:
The proper solution would be to adapt the public key as well. If the new public key needs to support fewer elements then the public key doesn't need to be created again and you can just drop the extra elements. And for creating a public key with more elements, new secret key needs to be created with the same seed as before so the signer is expected to preserve the seed or the key adaptation can be done only for shorter keys

Reply:
> The key adaptation can be done only for shorter keys
Unfortunately, `hash_to_field` called with different messages count will produce totally different results for identical seed, and elements on the same position won't be equal, i.e. `hash_to_field(seed, 1)[0] != hash_to_field(seed, 2)[0]`. Thus, `SecretKey::from_seed(seed, 10) != SecretKey::from_seed(seed, 11).adapt(10)`. 
> And for creating a public key with more elements, new secret key needs to be created with the same seed 
This's what I do already in tests (except for the seed being random, but it can be deterministic for sure).

Reply:
> Unfortunately, hash_to_field called with different messages count will produce totally different results for identical seed,

I didn't know that. So we can't get around that unless we deviate from the standard `hash_to_field` as the spec mandates to hash the number of output elements as well. We should only support the shortening of keys then.

Reply:
Shortening won't work too as `hash_to_field(seed, 5) != hash_to_field(seed, 10)[..5]` => `SecretKey::from_seed(seed, 5) != SecretKey::from_seed(seed, 10).adapt(5)`, so we would need to fully recreate `y` from seed anyway in both cases.

Reply:
You don't need to call `hash_to_field` to shorten, just drop the extra elements from secret key while signing. Similar for verifying.

Regarding supporting adaptation for longer keys, we can change the secret key generation process to call  `hash_to_field` multiple times rather than once. So to generate secret key for `n` messages, rather than calling it like 
`hash_to_field(seed, n)`, we can call like

```
sk = []
for i in 0..n {
    sk.push(hash_to_field(concat(seed,i), 1)); // 
}
```

Its more expensive but this operation will be done infrequently.

Reply:
> 5 messages were signed and sent to the second recipient but the public key was generated later from a new secret key created from supplied seed and 5 message count

The original keys created by the signer and verifier should always be available. So this is how I imagine the process:
1. Signer generates a keypair `sk` and `pk` supporting 10 messages.
2. When a key needs to be shortened to 5 messages, the signer/receiver will create `sk_5` by dropping 5 elements from `sk`. Similarly verifier creates `pk_5` from `pk`.
3. Next time when the signer needs to sign 6 messages, he again uses `sk` to create `sk_6`. Similar for the verifier.

> I'm totally fine with the idea to force secret key recreation.

Ok, then let's do that.